tests/compat/test_requests.py

```diff
@@ -0,0 +1,2285 @@
+            "E8A3BE87-9E3F-4620-8858-95478E385B5B",
+            "EA770032-DA4D-4D84-8CE9-29C6D910BF1E",
+            "exactly-------------sixty-----------three------------characters",
+        )
+        r = requests.Request("GET", url)
+        assert str(r.url) == url
+
+    def test_header_keys_are_native(self, httpbin):
+        headers = {"unicode": "blah", "byte".encode("ascii"): "blah"}
+        r = requests.Request("GET", httpbin("get"), headers=headers)
+
+        # This is testing that they are builtin strings. A bit weird, but there
+        # we go.
+        assert "unicode" in r.headers.keys()
+        assert "byte" in r.headers.keys()
+
+    def test_header_validation(self, httpbin):
+        """Ensure prepare_headers regex isn't flagging valid header contents."""
+        headers_ok = {
+            "foo": "bar baz qux",
+            "bar": "fbbq".encode("utf8"),
+            "baz": "",
+            "qux": "1",
+        }
+        r = requests.get(httpbin("get"), headers=headers_ok)
+        assert r.request.headers["foo"] == headers_ok["foo"]
+
+    @pytest.mark.httpx_bug
+    @pytest.mark.xfail(
+        reason=(
+            "HTTPX doesn't validate header values "
+            "(and does not have an equivalent for Requests' InvalidHeader exception)"
+        )
+    )
+    def test_header_value_not_str(self, httpbin):
+        """Ensure the header value is of type string or bytes as
+        per discussion in GH issue #3386
+        """
+        headers_int = {"foo": 3}
+        headers_dict = {"bar": {"foo": "bar"}}
+        headers_list = {"baz": ["foo", "bar"]}
+
+        # Test for int
+        with pytest.raises(InvalidHeader) as excinfo:
+            r = requests.get(httpbin("get"), headers=headers_int)
+        assert "foo" in str(excinfo.value)
+        # Test for dict
+        with pytest.raises(InvalidHeader) as excinfo:
+            r = requests.get(httpbin("get"), headers=headers_dict)
+        assert "bar" in str(excinfo.value)
+        # Test for list
+        with pytest.raises(InvalidHeader) as excinfo:
+            r = requests.get(httpbin("get"), headers=headers_list)
+        assert "baz" in str(excinfo.value)
+
+    @pytest.mark.httpx_bug
+    @pytest.mark.xfail(
+        reason=(
+            "HTTPX doesn't validate header values "
+            "(and does not have an equivalent for Requests' InvalidHeader exception)"
+        )
+    )
+    def test_header_no_return_chars(self, httpbin):
+        """Ensure that a header containing return character sequences raise an
+        exception. Otherwise, multiple headers are created from single string.
+        """
+        headers_ret = {"foo": "bar\r\nbaz: qux"}
+        headers_lf = {"foo": "bar\nbaz: qux"}
+        headers_cr = {"foo": "bar\rbaz: qux"}
+
+        # Test for newline
+        with pytest.raises(InvalidHeader):
+            r = requests.get(httpbin("get"), headers=headers_ret)
+        # Test 
```

Comment:
This is a bug, since it's possible to know the size of the request body via .tell() we should be setting a content length header.